### ISSUE TITLE
Better error message  for "Not a valid let declaration"

### ISSUE THREAD

**Issue by mechvel** [agda/agda#4586]
Agda reports "invalid `let' declaration" with copying the whole ``let`` construction and  without explaining a reason.
Could it mention  a reason in some cases? For example, for the following code, this may be 
the report  of "``where`` used inside ``let`` ...".
```
let ... 
     eq = begin ...        ≈⟨ ... ⟩  
                       ...
            ∎
           where open EqReasoning ...
in
 foo 
```

**Comment by andreasabel** [comment]
> Agda reports "invalid `let' declaration"

I grepped for that error and found nothing.  It is of great help to report problems as precisely as possible, since no one gains anything if developers waste time needlessly.

I could trigger:
```
Not a valid let-declaration
```
There are many reasons why something that parses as a `let` binding is later rejected, but in principle these are all syntax errors, and in the end, something isn't valid syntax, well, just because it isn't valid syntax.  We can try to point out likely mistakes, though, like having a `where` as above.

**Comment by andreasabel** [comment]
Playing with `let`, I stumbled over
```agda
test : Set₁
test = let ... | _ = Set
       in  Set
```
which raises an internal error:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs:1417
```

**Comment by andreasabel** [comment]
```agda
test : Set₁
test = let () = Set
       in  Set
```
This triggers a weird error:
```
Set₁ should be empty, but that's not obvious to me
when checking that the pattern () has type Set₁
```
Who claimed that `Set₁` should be empty, stupid hen?

**Comment by andreasabel** [comment]
Similar nonsense:
```agda
import Agda.Builtin.Nat

test : Set₁
test = let 3 = Set
       in  Set
```
```
Cannot split on argument of non-datatype Set₁
when checking that the pattern 3 has type Set₁
```

**Comment by andreasabel** [comment]
More amusement:
```agda
test : Set₁
test = let .Set = Set
       in  Set

-- Set != x of type Set₁
-- when checking that the given dot pattern Set matches the inferred
-- value x
```

**Comment by jespercockx** [comment]
Some of these are not really specific to pattern matching `let`s, if you write `.Set` in a normal pattern you get the same error.

**Comment by andreasabel** [comment]
Yes.
But here I have a chance to give a better error, but excluding early the pattern forms that do not make sense here.  Fixed these instances in faca04a.

What is a bit worrying is the `x` in
```
when checking that the given dot pattern Set matches the inferred value x
```
There is no `x` in sight here, how could it have been inferred?

**Comment by jespercockx** [comment]
The `x` is the name of the argument that Agda makes up before doing any case splitting (it would've used any user-provided name, but here the user provided nothing so it falls back to `x`). To check the dot pattern, Agda simply checks that the value of the pattern after all case splitting is equal to the expression behind the dot. Since there is no pattern matching here, the value of the pattern is simply `x`, and Agda complains that it is not equal to `Set`. So everything is working as expected, it's just the error message that's not very intuitive.

The fact that Agda makes up names for variables that are out of scope to the user should not be surprising. However, perhaps it would be possible to give a better error message for a dot pattern in case the internal pattern was not refined at all.
